// @okta/okta-auth-js/cjs/idx/remediators/Base/AuthenticatorData.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.AuthenticatorData = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Remediator = require("./Remediator");

// Base class - DO NOT expose static remediationName
class AuthenticatorData extends _Remediator.Remediator {
  constructor(remediation, values = {}) {
    super(remediation, values); // Unify authenticator input type

    (0, _defineProperty2.default)(this, "map", {
      'authenticator': []
    });
    const {
      authenticators
    } = this.values;
    const authenticatorType = this.getRelatesToType();
    const authenticator = authenticators === null || authenticators === void 0 ? void 0 : authenticators.find(authenticator => authenticator.type === authenticatorType);

    if (authenticator) {
      // map
      this.values.authenticators = authenticators.map(authenticator => {
        if (authenticatorType === authenticator.type) {
          return this.mapAuthenticatorFromValues(authenticator);
        }

        return authenticator;
      });
    } else {
      // add
      this.values.authenticators = [...authenticators, this.mapAuthenticatorFromValues()];
    }
  }

  getNextStep() {
    const common = super.getNextStep();
    const options = this.getMethodTypes();
    return { ...common,
      ...(options && {
        options
      })
    };
  } // Grab authenticator from authenticators list


  getAuthenticatorFromValues() {
    if (!this.values.authenticators) {
      return null;
    }

    const authenticatorType = this.getRelatesToType();
    const authenticator = this.values.authenticators.find(authenticator => authenticator.type === authenticatorType);
    return authenticator;
  }

  mapAuthenticatorFromValues(authenticator) {
    // add methodType to authenticator if it exists in values
    const type = this.getRelatesToType();
    const {
      methodType
    } = this.values;
    return {
      type,
      ...(authenticator && authenticator),
      ...(methodType && {
        methodType
      })
    };
  }

  getAuthenticatorFromRemediation() {
    const authenticator = this.remediation.value.find(({
      name
    }) => name === 'authenticator');
    return authenticator;
  }

  getMethodTypes() {
    var _authenticator$form$v;

    const authenticator = this.getAuthenticatorFromRemediation();
    return (_authenticator$form$v = authenticator.form.value.find(({
      name
    }) => name === 'methodType')) === null || _authenticator$form$v === void 0 ? void 0 : _authenticator$form$v.options;
  }

}

exports.AuthenticatorData = AuthenticatorData;
