// @okta/okta-auth-js/cjs/idx/remediators/Base/VerifyAuthenticator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.VerifyAuthenticator = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Remediator = require("./Remediator");

// Base class - DO NOT expose static remediationName
class VerifyAuthenticator extends _Remediator.Remediator {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "map", {
      'credentials': []
    });
  }

  canRemediate() {
    const challengeType = this.getRelatesToType();

    if (this.values.verificationCode && ['email', 'phone'].includes(challengeType)) {
      return true;
    }

    if (this.values.password && challengeType === 'password') {
      return true;
    }

    return false;
  }

  mapCredentials() {
    return {
      passcode: this.values.verificationCode || this.values.password
    };
  }

  getInputCredentials(input) {
    const challengeType = this.getRelatesToType();
    const name = challengeType === 'password' ? 'password' : 'verificationCode';
    return { ...input.form.value[0],
      name,
      type: 'string',
      required: input.required
    };
  }

}

exports.VerifyAuthenticator = VerifyAuthenticator;
(0, _defineProperty2.default)(VerifyAuthenticator, "remediationName", 'challenge-authenticator');
