// @okta/okta-auth-js/cjs/idx/types/index.js
"use strict";

Object.defineProperty(exports, "IdxMessage", {
  enumerable: true,
  get: function () {
    return _idxJs.IdxMessage;
  }
});
Object.defineProperty(exports, "AuthenticationOptions", {
  enumerable: true,
  get: function () {
    return _authenticate.AuthenticationOptions;
  }
});
Object.defineProperty(exports, "RegistrationOptions", {
  enumerable: true,
  get: function () {
    return _register.RegistrationOptions;
  }
});
Object.defineProperty(exports, "PasswordRecoveryOptions", {
  enumerable: true,
  get: function () {
    return _recoverPassword.PasswordRecoveryOptions;
  }
});
Object.defineProperty(exports, "CancelOptions", {
  enumerable: true,
  get: function () {
    return _cancel.CancelOptions;
  }
});
exports.IdxFeature = exports.IdxStatus = void 0;

var _idxJs = require("./idx-js");

var _authenticate = require("../authenticate");

var _register = require("../register");

var _recoverPassword = require("../recoverPassword");

var _cancel = require("../cancel");

let IdxStatus;
exports.IdxStatus = IdxStatus;

(function (IdxStatus) {
  IdxStatus[IdxStatus["SUCCESS"] = 0] = "SUCCESS";
  IdxStatus[IdxStatus["PENDING"] = 1] = "PENDING";
  IdxStatus[IdxStatus["FAILURE"] = 2] = "FAILURE";
  IdxStatus[IdxStatus["TERMINAL"] = 3] = "TERMINAL";
  IdxStatus[IdxStatus["CANCELED"] = 4] = "CANCELED";
})(IdxStatus || (exports.IdxStatus = IdxStatus = {}));

let IdxFeature;
exports.IdxFeature = IdxFeature;

(function (IdxFeature) {
  IdxFeature[IdxFeature["PASSWORD_RECOVERY"] = 0] = "PASSWORD_RECOVERY";
  IdxFeature[IdxFeature["REGISTRATION"] = 1] = "REGISTRATION";
  IdxFeature[IdxFeature["SOCIAL_IDP"] = 2] = "SOCIAL_IDP";
})(IdxFeature || (exports.IdxFeature = IdxFeature = {}));
